Scripts/Modules/SnappingModule/SnappingModule.cs

```diff
@@ -0,0 +1,709 @@
+		{
+			if (objects.Length == 0)
+				return false;
+
+			if (snappingEnabled && manipulatorSnapping)
+			{
+				var state = GetSnappingState(rayOrigin, objects[0], position, rotation);
+
+				state.currentPosition += delta;
+				var targetPosition = state.currentPosition;
+				var targetRotation = state.startRotation;
+
+				var camera = CameraUtils.GetMainCamera();
+				var breakScale = Vector3.Distance(camera.transform.position, targetPosition);
+
+				if (surfaceSnapping && PerformManipulatorSurfaceSnapping(rayOrigin, ref position, ref rotation, targetPosition, state, targetRotation, breakScale * k_ManipulatorSurfaceSnapBreakDist))
+					return true;
+				if (localOnly)
+				{
+					if (groundSnapping && PerformGroundSnapping(ref position, ref rotation, targetPosition, targetRotation, state, breakScale * k_ManipulatorGroundSnapMin, breakScale * k_ManipulatorGroundSnapMax))
+						return true;
+				}
+				else
+				{
+					var groundPlane = new Plane(Vector3.up, k_GroundHeight);
+					var origin = rayOrigin.position;
+					var direction = rayOrigin.forward;
+					var ray = new Ray(origin, direction);
+					var raycastDistance = k_GroundSnappingMaxRayLength * this.GetViewerScale();
+					float distance;
+					if (groundPlane.Raycast(ray, out distance) && distance <= raycastDistance)
+					{
+						state.groundSnapping = true;
+
+						position = origin + direction * distance;
+
+						if (snapRotation)
+							rotation = Quaternion.LookRotation(Vector3.up, targetRotation * Vector3.back) * Quaternion.AngleAxis(90, Vector3.right);
+					}
+					else
+					{
+						state.groundSnapping = false;
+						position = targetPosition;
+						rotation = targetRotation;
+					}
+				}
+			}
+
+			position += delta;
+
+			return false;
+		}
+
+		public bool DirectSnapping(Transform rayOrigin, GameObject go, ref Vector3 position, ref Quaternion rotation, Vector3 targetPosition, Quaternion targetRotation)
+		{
+			if (snappingEnabled && directSnapping)
+			{
+				var state = GetSnappingState(rayOrigin, go, position, rotation);
+
+				state.currentPosition = targetPosition;
+
+				var viewerScale = this.GetViewerScale();
+				var breakScale = viewerScale;
+				var breakDistance = breakScale * k_DirectSurfaceSnapBreakDist;
+
+				if (surfaceSnapping && PerformDirectSurfaceSnapping(ref position, ref rotation, targetPosition, state, targetRotation, breakDistance))
+					return true;
+
+				if (groundSnapping && PerformGroundSnapping(ref position, ref rotation, targetPosition, targetRotation, state, breakScale * k_DirectGroundSnapMin, breakScale * k_DirectGroundSnapMax))
+					return true;
+			}
+
+			position = targetPosition;
+			rotation = targetRotation;
+
+			return false;
+		}
+
+		bool PerformManipulatorSurfaceSnapping(Transform rayOrigin, ref Vector3 position, ref Quaternion rotation, Vector3 targetPosition, SnappingState state, Quaternion targetRotation, float breakDistance)
```

Comment:
Instead of Perform is there a verb we can use instead? ManipulatorSnapToSurface?